File: internal/memory/invariants.go
package memory

import (
	"fmt"
	"strings"
	"time"
	"unicode/utf8"
)

type Invariant struct {
	ID        string    `json:"id"`
	Category  string    `json:"category"`
	Title     string    `json:"title"`
	Rule      string    `json:"rule"`
	Reason    string    `json:"reason"`
	Status    string    `json:"status"` // active, disabled, proposed, rejected
	Version   int       `json:"version"`
	UpdatedAt time.Time `json:"updatedAt"`
}
type InvariantEvent struct {
	Action string    `json:"action"`
	Rule   Invariant `json:"rule"`
	At     time.Time `json:"at"`
}
type InvariantSet struct {
	Version int              `json:"version"`
	Items   []Invariant      `json:"items"`
	Events  []InvariantEvent `json:"events"`
}
type InvariantCommand struct {
	TaskID    string    `json:"taskId"`
	ProfileID string    `json:"profileId"`
	Version   int       `json:"version"`
	Action    string    `json:"action"`
	Rule      Invariant `json:"rule"`
}

func emptyInvariants() InvariantSet {
	return InvariantSet{Version: 1, Items: []Invariant{}, Events: []InvariantEvent{}}
}
func (s InvariantSet) Active() []Invariant {
	result := []Invariant{}
	for _, r := range s.Items {
		if r.Status == "active" {
			result = append(result, r)
		}
	}
	return result
}
func validInvariant(r *Invariant) error {
	r.Title = strings.TrimSpace(r.Title)
	r.Rule = strings.TrimSpace(r.Rule)
	r.Reason = strings.TrimSpace(r.Reason)
	if r.Category != "architecture" && r.Category != "decision" && r.Category != "stack" && r.Category != "business" && r.Category != "other" {
		return fmt.Errorf("%w: выберите категорию", ErrInvalid)
	}
	if r.Title == "" || r.Rule == "" || utf8.RuneCountInString(r.Title) > 100 || utf8.RuneCountInString(r.Rule) > 2000 || utf8.RuneCountInString(r.Reason) > 1000 {
		return fmt.Errorf("%w: название до 100, правило до 2000, обоснование до 1000 символов", ErrInvalid)
	}
	return nil
}
func (s *Store) UpdateInvariants(owner string, cmd InvariantCommand) (State, error) {
	return s.update(owner, func(state *State) error {
		set := &state.Invariants
		if state.Task.ID != cmd.TaskID || set.Version != cmd.Version {
			return ErrConflict
		}
		index := -1
		for i, r := range set.Items {
			if r.ID == cmd.Rule.ID {
				index = i
				break
			}
		}
		r := cmd.Rule
		switch cmd.Action {
		case "create":
			if cmd.Rule.ID != "" {
				return ErrInvalid
			}
			if len(set.Items) >= 50 {
				return fmt.Errorf("%w: не более 50 правил на задачу", ErrInvalid)
			}
			if err := validInvariant(&r); err != nil {
				return err
			}
			id, err := newID()
			if err != nil {
				return err
			}
			r.ID = id
			r.Status = "active"
			r.Version = 1
		case "edit", "activate", "disable", "reject":
			if index < 0 {
				return ErrConflict
			}
			old := set.Items[index]
			if cmd.Action == "edit" {
				if err := validInvariant(&r); err != nil {
					return err
				}
				r.Status = old.Status
			} else {
				r = old
				switch cmd.Action {
				case "activate":
					if old.Status != "disabled" && old.Status != "proposed" {
						return ErrConflict
					}
					r.Status = "active"
				case "disable":
					if old.Status != "active" {
						return ErrConflict
					}
					r.Status = "disabled"
				case "reject":
					if old.Status != "proposed" {
						return ErrConflict
					}
					r.Status = "rejected"
				}
			}
			r.Version = old.Version + 1
		default:
			return ErrInvalid
		}
		r.UpdatedAt = time.Now().UTC()
		if index < 0 {
			set.Items = append(set.Items, r)
		} else {
			set.Items[index] = r
		}
		set.Version++
		set.Events = append(set.Events, InvariantEvent{Action: cmd.Action, Rule: r, At: r.UpdatedAt})
		if len(set.Events) > 200 {
			set.Events = set.Events[len(set.Events)-200:]
		}
		// A changed rule set invalidates old chat requests and workflow suggestions.
		state.Task.Workflow.Version++
		state.Task.Workflow.Proposal = nil
		return nil
	})
}
func (s *Store) ProposeInvariants(owner, taskID string, proposals []Invariant) (State, error) {
	return s.update(owner, func(state *State) error {
		if state.Task.ID != taskID {
			return ErrConflict
		}
		if len(proposals) > 3 {
			return ErrInvalid
		}
		for _, r := range proposals {
			if err := validInvariant(&r); err != nil {
				return err
			}
			duplicate := false
			for _, old := range state.Invariants.Items {
				if old.Rule == r.Rule || old.Title == r.Title {
					duplicate = true
				}
			}
			if duplicate {
				continue
			}
			if len(state.Invariants.Items) >= 50 {
				break
			}
			id, err := newID()
			if err != nil {
				return err
			}
			r.ID = id
			r.Status = "proposed"
			r.Version = 1
			r.UpdatedAt = time.Now().UTC()
			state.Invariants.Items = append(state.Invariants.Items, r)
			state.Invariants.Version++
		}
		return nil
	})
}

File: internal/web/static/invariants.js
(() => {
  const categories = { architecture: "Архитектура", decision: "Техническое решение", stack: "Стек", business: "Бизнес-правило", other: "Другое" };
  const statuses = { active: "Активен", disabled: "Отключён", proposed: "На подтверждении", rejected: "Отклонён" };
  function node(tag, text, cls) { const el = document.createElement(tag); if (text !== undefined) el.textContent = text; if (cls) el.className = cls; return el; }
  function button(text, action) { const el = node("button", text, "memory-button"); el.type = "button"; el.addEventListener("click", action); return el; }
  function editor(rule, save) {
    const form = node("form", undefined, "invariant-form");
    const field = (name, label, type, value, options) => {
      const wrap = node("label", label); const input = node(type); input.name = name;
      if (options) for (const [key, text] of Object.entries(options)) { const opt = node("option", text); opt.value = key; input.append(opt); }
      input.value = value || ""; wrap.append(input); form.append(wrap); return input;
    };
    field("category", "Категория", "select", rule.category || "stack", categories);
    const title = field("title", "Название", "input", rule.title); title.required = true; title.maxLength = 100;
    const text = field("rule", "Правило", "textarea", rule.rule); text.required = true; text.maxLength = 2000; text.rows = 3;
    const reason = field("reason", "Обоснование", "textarea", rule.reason); reason.maxLength = 1000; reason.rows = 2;
    const submit = node("button", rule.id ? "Сохранить изменения" : "Добавить активный инвариант", "memory-button memory-primary"); submit.type = "submit"; form.append(submit);
    form.addEventListener("submit", event => { event.preventDefault(); save({ ...rule, ...Object.fromEntries(new FormData(form)) }); });
    return form;
  }
  function render(state, mutate) {
    const set = state.invariants;
    const command = (action, rule) => mutate("/api/tasks/invariants", { taskId: state.task.id, profileId: window.CodexProfiles.activeID(), version: set.version, action, rule });
    const list = document.querySelector("#invariants-list"); list.replaceChildren();
    for (const rule of set.items || []) {
      const card = node("article", undefined, "memory-entry invariant-card");
      card.append(node("span", `${statuses[rule.status]} · ${categories[rule.category]} · v${rule.version}`, "task-badge"), node("h3", rule.title), node("p", rule.rule));
      if (rule.reason) card.append(node("p", rule.reason, "memory-reason"));
      const actions = node("div", undefined, "memory-actions");
      if (rule.status === "active") actions.append(button("Отключить", () => command("disable", rule)));
      if (rule.status === "disabled" || rule.status === "proposed") actions.append(button(rule.status === "proposed" ? "Подтвердить инвариант" : "Активировать", () => command("activate", rule)));
      if (rule.status === "proposed") actions.append(button("Отклонить", () => command("reject", rule)));
      card.append(actions);
      if (rule.status !== "rejected") { const details = node("details", undefined, "memory-edit"); details.append(node("summary", "Редактировать"), editor(rule, updated => command("edit", updated))); card.append(details); }
      list.append(card);
    }
    if (!set.items?.length) list.append(node("p", "Правил пока нет. Добавьте ограничения или подтвердите предложение агента. Контроль текущего этапа задачи работает независимо от этого списка."));
    const log = node("details", undefined, "memory-edit"); log.append(node("summary", `История изменений (${set.events?.length || 0})`));
    const names = { create: "Добавлен", edit: "Изменён", activate: "Активирован", disable: "Отключён", reject: "Отклонён" };
    for (const event of [...(set.events || [])].reverse()) log.append(node("p", `${new Date(event.at).toLocaleString("ru-RU")} · ${names[event.action]} · ${event.rule.title} v${event.rule.version}: ${event.rule.rule}`));
    list.append(log);
    document.querySelector("#invariant-create-form").replaceChildren(editor({}, rule => command("create", rule)));
  }
  function snapshot(check) {
    const root = node("details", undefined, "profile-snapshot invariant-snapshot");
    const labels = { allow: "проверено", partial: "частичный конфликт", conflict: "конфликт запроса", clarify: "нужно уточнение", rules_conflict: "противоречие правил", unavailable: "проверка недоступна", blocked: "решение заблокировано" };
    root.append(node("summary", `Учтённые ограничения · ${labels[check.status] || check.status}`), node("p", check.explanation));
    for (const rule of check.rules || []) root.append(node("p", `${(check.conflictingIds || []).includes(rule.id) ? "Конфликт · " : ""}${rule.title} v${rule.version}: ${rule.rule}`));
    return root;
  }
  window.CodexInvariants = Object.freeze({ render, snapshot });
})();
